[Enhancement] Support restore/rollback sync during conversion (1/3)

danielhumanmod opened the pull request:
## *Important Read*
- Aim to address #40 

## What is the purpose of the pull request

Previously, if a rollback/restore occurred in the source table, XTable would reflect it as file changes (added or deleted) in the target table. In this PR, we aim to improve this by issuing a rollback command in the target tables, ensuring more consistent histories between the source and target. This approach is also more efficient, as it allows us to restore directly to a specific version/snapshot instead of computing a large diff against the table’s current state.

This is the first part of this enhancement (1/3), focusing primarily on detecting whether a rollback/restore occurred in the source table and verifying if the corresponding commit exists in the target table.

## Brief change log

- *Add a source identifier in target transaction*
- *[Source] Detect rollback and get the rollbackSnapshot from source table*
- *[Target] Verify if the target table contains the corresponding rollback commit*

## Verify this pull request

This pull request is already covered by existing tests, all existing tests should pass

danielhumanmod commented on xtable-api/src/main/java/org/apache/xtable/spi/extractor/ExtractFromSource.java:
```diff
@@ -47,9 +49,20 @@ public IncrementalTableChanges extractTableChanges(
         commitsBacklog.getCommitsToProcess().stream()
             .map(conversionSource::getTableChangeForCommit)
```
> I think that we'll want the identifier on the commit level, right?


Thanks for the response @the-other-tim-brown. 

Yes, ideally, every commit in source table can directly map to one in target table. However, based on my understanding of how XTable works, this isn’t guaranteed. Instead, the mapping (Source -> Target) is a N:1 mapping, which means:
- Every commit in the target table has a corresponding mapping in the source table.
- Not every commit in the source table has a one-to-one mapping in the target table.

The reason is, between each sync(), there could be multiple changes on source, and all these changes will sync as only one commit in target, just like this example
```
Iceberg (Source)          Delta (Target)  
┌────────────┐      ┌─────────────────────┐
│ Snapshot 0 │ ◀  ▶ │ Version 0 (Synced)  │  
│ Snapshot 1 │      │                     │  
│ Snapshot 2 │      │                     │  
│ Snapshot 3 │      │                     │  
│ Snapshot 4 │      │                     │  
│ Snapshot 5 │ ◀  ▶ │ Version 1 (Synced)  │
└────────────┘      └─────────────────────┘  
```
Given this, I’ve chosen to use the information from the latest commit in the source table as the source identifier.
But my understanding might be wrong, please feel free to correct me if I’ve missed something :)

the-other-tim-brown replied:
This is true for snapshot sync but with incremental sync, there are multiple commits all synced to the target as their own commits

danielhumanmod replied:
Thanks for pointing this out—it’s an important detail! I’ll investigate further into the ability to track commit-level information in Iceberg and Delta. I will get back to you once I figure it out

the-other-tim-brown commented:
@danielhumanmod Thanks for looking into it. I think being able to store this at a commit level better aligns with my our intentions so it would be great to fix this.

danielhumanmod commented on xtable-api/src/test/java/org/apache/xtable/spi/sync/TestTableFormatSync.java:
TL;DR: Add the source identifier to snapshots or incremental changes in the test

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/spi/sync/ConversionTarget.java:
```diff
@@ -90,4 +91,7 @@ public interface ConversionTarget {
   /** Initializes the client with provided configuration */
   void init(TargetTable targetTable, Configuration configuration);
+
+  /** Return the commit identifier from target table */
```
Can you be more specific here? I am assuming this is the latest commit identifier but it should be more specific if the table can have more than one over time.

danielhumanmod replied:
> Can you be more specific here? I am assuming this is the latest commit identifier but it should be more specific if the table can have more than one over time.

Do you mean the snapshot sync scenario? Because In incremental sync, we will ensure target and source commit has 1:1 mapping. However, in snapshot sync, we will use the commit id of current (latest) snapshot even though the changes might contains more than one commit.

danielhumanmod replied:
To ensure we’re aligned, the context of this function is:

It is designed to retrieve the corresponding target commit based on a given source identifier during a sync operation. Here’s the scenario:

Assume:
- Last completed sync (1 snapshot, 1 incremental) is source commit 3
- We’re starting the new sync process from source commit 4

Source table commit history:
	•	1 (UPDATE)
	•	2 (UPDATE)
	•	3 (UPDATE)
	•	4 (ROLLBACK to 2)
	•	5 (UPDATE)
	•	6 (UPDATE)
Target table commit history (mapped by source identifiers):
	•	1 (mapped to source id 2)
	•	2 (mapped to source id 3)

When syncing the ROLLBACK operation (source commit 4), we need to identify the corresponding target commit that aligns with the source identifier 2 (which is 1 in target).

the-other-tim-brown replied:
That makes sense to me but the java doc can be a bit more clear. It should also include context on when the Optional will be empty

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/spi/sync/ConversionTarget.java:
```diff
@@ -76,8 +76,9 @@ public interface ConversionTarget {
+   * @param sourceIdentifier
    */
-  void beginSync(InternalTable table);
+  void beginSync(InternalTable table, String sourceIdentifier);
```
I am wondering if it makes sense to attach the source format as part of the identifier. Then we can also have some metadata about what the writer format was.

danielhumanmod replied:
> I am wondering if it makes sense to attach the source format as part of the identifier. Then we can also have some metadata about what the writer format was.

Good idea. I create a `SourceMetadata` as a commit-level information attached in target table commit so that It would be much easier if we want add more information in the future

danielhumanmod commented on xtable-core/src/main/java/org/apache/xtable/delta/DeltaConversionTarget.java:
```diff
@@ -215,13 +224,64 @@ public String getTableFormat() {
   }
 
+  @Override
+  public Optional<String> getTargetCommitIdentifier(String sourceIdentifier) {
```
> If there is no commit Identifier in the latest commit, should this just return an empty Optional?

Yes, if we can not find the corresponding target commit, we will return an empty Optional

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/metadata/TableSyncMetadata.java:
```diff
@@ -51,6 +51,8 @@ public class TableSyncMetadata {
   /** Property name for the XTABLE metadata in the table metadata/properties */
   public static final String XTABLE_METADATA = "XTABLE_METADATA";
 
+  public static final String XTABLE_SOURCE_METADATA = "XTABLE_SOURCE_METADATA";
```
Instead of having a separate metadata, is it possible to embed this within the metadata object that we have today?

danielhumanmod replied:
Good catch! The current metadata is table-level information, while the new one is commit-level, but it is true that we can reuse this key.

danielhumanmod replied:
And maybe we could consider moving the source format to table-level information and leaving the source identifier at the commit level to reduce redundancy.

danielhumanmod replied:
Ah I see your point, that's a good way. I’ll get started on it.

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/delta/DeltaConversionTarget.java:
```diff
@@ -264,7 +330,8 @@ private void commitTransaction() {
           actions,
-          new DeltaOperations.Update(Option.apply(Literal.fromObject("xtable-delta-sync"))));
+          new DeltaOperations.Update(Option.apply(Literal.fromObject("xtable-delta-sync"))),
+          ScalaUtils.convertJavaMapToScala(getCommitTags()));
```
What is the difference between putting this info in tags compared to the metadata we are currently using? Should we consolidate?

danielhumanmod replied:
Here `xtable-delta-sync` is the operation, and the tags is more recommended to store the metadata as k-v pairs

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaSync.java:
```diff
@@ -159,16 +161,22 @@ public void testCreateSnapshotControlFlow() throws Exception {
+    Optional<String> targetIdentifier2 = conversionTarget.getTargetCommitIdentifier("1");
     validateDeltaTable(basePath, new HashSet<>(Arrays.asList(dataFile2, dataFile3)), null);
+    assertTrue(targetIdentifier2.isPresent());
+    assertEquals("1", targetIdentifier2.get());
```
Let's have 1 test where we return an empty option as well

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -316,6 +323,53 @@ public String getTableFormat() {
+    return getTargetCommitIdentifier(sourceIdentifier, metaClient.get());
+  }
+
+  public Optional<String> getTargetCommitIdentifier(
```
Can this be made private or at least package private?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -316,6 +323,53 @@ public String getTableFormat() {
+
+  public Optional<String> getTargetCommitIdentifier(
+      String sourceIdentifier, HoodieTableMetaClient metaClient) {
+    long sourceIdentifierVal = Long.parseLong(sourceIdentifier);
```
This is assuming that all sources will have identifiers that can be mapped to longs, that may not be true in the future. Can we keep this as strings throughout?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -316,6 +323,53 @@ public String getTableFormat() {
+      String sourceIdentifier, HoodieTableMetaClient metaClient) {
+    long sourceIdentifierVal = Long.parseLong(sourceIdentifier);
+
+    HoodieTimeline completedTimeline = metaClient.getActiveTimeline().filterCompletedInstants();
```
We can filter this to the commits timeline (avoids looking at cleaner or other commit types) with `getCommitsTimeline`

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -316,6 +323,53 @@ public String getTableFormat() {
+
+    for (HoodieInstant instant : completedTimeline.getInstants()) {
+      try {
+        Option<byte[]> instantDetails = metaClient.getActiveTimeline().getInstantDetails(instant);
```
There is a similar block of code above (line 300) that also handles parsing of the "replace commit" type which is what we'll be writing from XTable. Can you extract that into a common helper method to avoid repeated code?

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -316,6 +323,53 @@ public String getTableFormat() {
+        SourceMetadata sourceMetadata = SourceMetadata.fromJson(sourceMetadataJson);
+
+        if (sourceIdentifier.equals(sourceMetadata.getSourceIdentifier())) {
+          return Optional.of(sourceIdentifier);
```
Should we return the identifier of the target?

danielhumanmod commented on xtable-api/src/main/java/org/apache/xtable/spi/sync/TableFormatSync.java:
```diff
@@ -149,19 +151,26 @@ private SyncResult getSyncResult(
+      String sourceIdentifier) {
     // initialize the sync
     conversionTarget.beginSync(tableState);
+    // Persist the latest commit time in table properties for incremental syncs
```
Here We need to move the Metadata set operation earlier because it will be required during the `sync()` operation in Iceberg (Delta and Hudi only need it in `completeSync()`

danielhumanmod commented on xtable-core/src/main/java/org/apache/xtable/iceberg/IcebergDataFileUpdatesSync.java:
```diff
@@ -60,32 +62,37 @@ public void applySnapshot(
     OverwriteFiles overwriteFiles = transaction.newOverwrite();
     filesAdded.forEach(f -> overwriteFiles.addFile(getDataFile(partitionSpec, schema, f)));
     filesRemoved.forEach(overwriteFiles::deleteFile);
+    overwriteFiles.set(TableSyncMetadata.XTABLE_METADATA, metadata.toJson());
```
Attach XTable metadata as Iceberg snapshot summary (commit level)

danielhumanmod commented:
Thanks for the suggestions @the-other-tim-brown ! I have updated the code in latest commit, including changes:
- All XTABLE_METADATA are commit level across 3 table format
- Adjust `getTargetCommitIdentifier` method
- Refactor metadata retrieval logic in `HudiConversionTarget'
- Added test cases for empty target commit identifier.

Appreciate any further suggestion in advanced!

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/hudi/HudiConversionTarget.java:
```diff
@@ -287,40 +287,81 @@ public Optional<TableSyncMetadata> getTableMetadata() {
+
+        TableSyncMetadata metadata = optionalMetadata.get();
+        if (sourceIdentifier.equals(metadata.getSourceIdentifier())) {
+          return Optional.of(String.valueOf(instant.getTimestamp()));
```
`getTimestamp` is already returning a string, can we remove the `String.valueOf`?

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/model/TableChange.java:
```diff
@@ -36,4 +36,7 @@ public class TableChange {
   InternalTable tableAsOfChange;
+
+  // Commit identifier in source table
+  @Builder.Default String sourceIdentifier = "";
```
I think this may be safer without a default and require non-null, what do you think?

Similar note for `InternalSnapshot`

the-other-tim-brown reviewed (commented):
@danielhumanmod apologies for the delay on this review. I had lost track of this change. One concern I have is how do we migrate existing tables. Should we support the old way of reading the commit metadata as a fallback so we can easily migrate users without requiring them to do a full snapshot sync?

danielhumanmod commented:
Sorry for the delayed response @the-other-tim-brown . I just returned from vacation with my family.

Thank you for your suggestions! You’re absolutely right—backward compatibility is important. In my latest commit, I’ve preserved the existing table-level metadata while also adding commit-level tracking to support future rollback features. Additionally, I’ve addressed the issues mentioned in your previous comments.

I appreciate any further feedback!

the-other-tim-brown commented:
@danielhumanmod I was out sick last week but I will be taking a look at this PR this week. Thank you for handling the compatibility concerns! We have also started a bi-weekly developer sync and slack if you are looking to connect with others working on the project https://xtable.apache.org/community/sync#bi-weekly-community-call

danielhumanmod commented:
> @danielhumanmod I was out sick last week but I will be taking a look at this PR this week. Thank you for handling the compatibility concerns! We have also started a bi-weekly developer sync and slack if you are looking to connect with others working on the project https://xtable.apache.org/community/sync#bi-weekly-community-call

Thanks, please take your time and take care of yourself, Tim! I’m really glad we have a community Slack and regular syncs now. I’ll join the next one if I am available! :)

the-other-tim-brown commented on xtable-api/src/main/java/org/apache/xtable/spi/sync/ConversionTarget.java:
```diff
@@ -90,4 +90,14 @@ public interface ConversionTarget {
   void init(TargetTable targetTable, Configuration configuration);
+
+  /**
+   * Retrieves the commit identifier from the target table that corresponds to a given source table
+   * commit identifier
```
Will these identifier's always be the same? if so, is it simpler to make this a boolean method?

danielhumanmod replied:
> Will these identifier's always be the same? if so, is it simpler to make this a boolean method?

Yes the source identifier stored in the target table remains unchanged. However, this API is designed to retrieve the corresponding target COMMIT based on an input source identifier (It would be the rollback source COMMIT in this feature). This allows us to initiate a rollback to the retrieved target COMMIT on the target table

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/iceberg/TestIcebergSync.java:
```diff
@@ -244,11 +244,21 @@ public void testCreateSnapshotControlFlow() throws Exception {
+    Optional<String> targetIdentifier1 =
+        conversionTarget.getTargetCommitIdentifier(snapshot1.getSourceIdentifier());
     validateIcebergTable(tableName, table1, Sets.newHashSet(dataFile1, dataFile2), null);
+    assertTrue(targetIdentifier1.isPresent());
```
Can we assert on the content of the Optional?

the-other-tim-brown commented on xtable-core/src/test/java/org/apache/xtable/hudi/ITHudiConversionTarget.java:
```diff
@@ -367,6 +382,20 @@ CONTEXT, getHoodieWriteConfig(metaClient).getMetadataConfig(), tableBasePath, tr
       assertColStats(hoodieBackedTableMetadata, partitionPath, fileName4);
     }
+    Optional<String> targetIdentifier3 = targetClient.getTargetCommitIdentifier("2", metaClient);
+    assertTrue(targetIdentifier3.isPresent());
```
Can we assert on the content of the Optional?

danielhumanmod commented on xtable-core/src/test/java/org/apache/xtable/hudi/ITHudiConversionTarget.java:
```diff
@@ -367,23 +380,113 @@ CONTEXT, getHoodieWriteConfig(metaClient).getMetadataConfig(), tableBasePath, tr
     assertEquals(
         2, metaClient.getArchivedTimeline().reload().filterCompletedInstants().countInstants());
   }
 
```
To make test clean, create a new test to verify source-target identifier mapping

danielhumanmod commented on xtable-core/src/test/java/org/apache/xtable/iceberg/TestIcebergSync.java:
```diff
@@ -681,10 +681,69 @@ public void testNestedFieldPartitioning() throws Exception {
         Expressions.equal(partitionField.getSourceField().getPath(), "value1"));
   }
 
```
Likewise, to make test clean, create a new test to verify source-target identifier mapping

danielhumanmod commented on xtable-core/src/test/java/org/apache/xtable/delta/TestDeltaSync.java:
```diff
@@ -367,6 +369,45 @@ public void testTimestampPartitioning(PartitionTransformType transformType) thro
             .contains(String.format("xtable_partition_col_%s_timestamp_field", transformType)));
   }
 
```
Likewise, to make test clean, create a new test to verify source-target identifier mapping

the-other-tim-brown commented on xtable-api/src/test/java/org/apache/xtable/model/metadata/TestTableSyncMetadata.java:
```diff
@@ -49,14 +49,17 @@ private static Stream<Arguments> provideMetadataAndJson() {
-            TableSyncMetadata.of(Instant.parse("2020-07-04T10:15:30.00Z"), null),
-            "{\"lastInstantSynced\":\"2020-07-04T10:15:30Z\",\"version\":0}"));
+            TableSyncMetadata.of(Instant.parse("2020-07-04T10:15:30.00Z"), null, "TEST", "0"),
+            "{\"lastInstantSynced\":\"2020-07-04T10:15:30Z\",\"version\":0,\"sourceTableFormat\":\"TEST\",\"sourceIdentifier\":\"0\"}"));
```
Can we have one test where the fields are not present to ensure that the parsing will not fail when reading from the previous version's metadata?

danielhumanmod commented on xtable-api/src/test/java/org/apache/xtable/model/metadata/TestTableSyncMetadata.java:
New data and test case added for backward compatibility

the-other-tim-brown commented on xtable-api/src/test/java/org/apache/xtable/model/metadata/TestTableSyncMetadata.java:
```diff
@@ -56,7 +59,24 @@ private static Stream<Arguments> provideMetadataAndJson() {
             TableSyncMetadata.of(Instant.parse("2020-07-04T10:15:30.00Z"), null),
-            "{\"lastInstantSynced\":\"2020-07-04T10:15:30Z\",\"version\":0}"));
+            "{\"lastInstantSynced\":\"2020-07-04T10:15:30Z\",\"version\":0}"),
+        // New version of metadata and JSON where two new fields are added
```
"new" will be relative in the future so we should mention the names of the fields added

the-other-tim-brown commented on xtable-core/src/main/java/org/apache/xtable/delta/DeltaConversionTarget.java:
```diff
@@ -258,10 +258,15 @@ public Optional<String> getTargetCommitIdentifier(String sourceIdentifier) {
         }
 
         TableSyncMetadata metadata = optionalMetadata.get();
+        // For backward compatability since the older version JSON might not contain new fields
```
I think this will be covered at line 270 so we don't need the extra condition right?

Is there a test case where we will have a `null` to cover this case?

the-other-tim-brown reviewed (approved):
Thanks @danielhumanmod! Can you squash down to a single commit?

danielhumanmod commented:
> Thanks @danielhumanmod! Can you squash down to a single commit?

Done! Thanks for all the feedback and suggestion during this PR @the-other-tim-brown , really appreciate it. When you have a moment, could you help merge it? :)
